File: squash-python/squash/v2/varint.py
"""
Unsigned Varint (LEB128) encoder/decoder for SQUASH v2.

Each byte uses 7 bits for data and 1 bit (MSB) as a continuation flag.
This is the same encoding used by Protocol Buffers.

Size characteristics:
    Values 0–127:       1 byte
    Values 128–16383:   2 bytes
    Values 16384–2097151: 3 bytes

Critical for SQUASH v2: Protobuf field tags for fields 1–15 encode as
single-byte Varints, making the most common fields extremely compact.
"""

from __future__ import annotations


class Varint:
    """LEB128 unsigned Varint codec."""

    @staticmethod
    def encode_zigzag(value: int) -> int:
        """ZigZag encode a signed integer into an unsigned integer."""
        return (value << 1) ^ (value >> 63)

    @staticmethod
    def decode_zigzag(value: int) -> int:
        """ZigZag decode an unsigned integer back to a signed integer."""
        return (value >> 1) ^ -(value & 1)

    @staticmethod
    def encode(value: int) -> bytes:
        """
        Encode an unsigned integer as a Varint byte sequence.

        Args:
            value: Non-negative integer to encode.

        Returns:
            bytes containing the Varint encoding.

        Raises:
            ValueError: If value is negative.
        """
        if value < 0:
            raise ValueError(f"Varint value must be non-negative, got {value}")

        result = bytearray()
        while True:
            byte = value & 0x7F
            value >>= 7
            if value != 0:
                byte |= 0x80
            result.append(byte)
            if value == 0:
                break

        return bytes(result)

    @staticmethod
    def decode(data: bytes | bytearray, offset: int = 0) -> tuple[int, int]:
        """
        Decode a Varint from a byte sequence.

        Args:
            data: Byte sequence to read from.
            offset: Starting position in the sequence.

        Returns:
            Tuple of (decoded_value, bytes_consumed).

        Raises:
            ValueError: If the varint is malformed or truncated.
        """
        result = 0
        shift = 0
        pos = offset

        while pos < len(data):
            byte = data[pos]
            result |= (byte & 0x7F) << shift
            pos += 1

            if byte & 0x80 == 0:
                return result, pos - offset

            shift += 7
            if shift >= 64:
                raise ValueError("Varint too long (more than 10 bytes)")

        raise ValueError(f"Unexpected end of varint at offset {offset}")

    @staticmethod
    def encode_tag(field_number: int, wire_type: int) -> bytes:
        """
        Encode a Protobuf field tag (field_number << 3 | wire_type).

        Args:
            field_number: 1-based field number.
            wire_type: Wire type constant (0, 1, 2, or 5).

        Returns:
            Varint-encoded tag bytes.
        """
        tag = (field_number << 3) | wire_type
        return Varint.encode(tag)

File: squash-core/src/commonMain/kotlin/com/squash/core/v2/wire/Varint.kt
package com.squash.core.v2.wire

/**
 * Unsigned Varint (LEB128) encoder/decoder.
 *
 * Varints are a compact encoding for unsigned integers used in Protocol Buffers.
 * Each byte uses 7 bits for data and 1 bit (MSB) as a continuation flag:
 *
 * ```
 * Value 300 (0x012C):
 *   Byte 1: 1_0101100  (0xAC) — MSB=1 means "more bytes follow"
 *   Byte 2: 0_0000010  (0x02) — MSB=0 means "final byte"
 * ```
 *
 * Size characteristics:
 * - Values 0–127: 1 byte
 * - Values 128–16383: 2 bytes
 * - Values 16384–2097151: 3 bytes
 *
 * This is critical for SQUASH v2: field tags 1–15 encode as single bytes,
 * making the 15 most common fields extremely space-efficient.
 */
object Varint {

    /**
     * Encodes an unsigned integer as a Varint byte sequence.
     *
     * @param value The unsigned value to encode (treated as unsigned).
     * @return ByteArray containing the Varint encoding.
     */
    fun encode(value: Long): ByteArray {
        if (value < 0) {
            // Negative values need 10 bytes in protobuf varint encoding
            return encodeFull(value)
        }

        val bytes = mutableListOf<Byte>()
        var remaining = value

        do {
            var byte = (remaining and 0x7F).toByte()
            remaining = remaining ushr 7
            if (remaining != 0L) {
                byte = (byte.toInt() or 0x80).toByte()
            }
            bytes.add(byte)
        } while (remaining != 0L)

        return bytes.toByteArray()
    }

    /**
     * Encodes an Int as a Varint.
     */
    fun encode(value: Int): ByteArray = encode(value.toLong())

    /**
     * Encodes an unsigned Int as a Varint.
     */
    fun encode(value: UInt): ByteArray = encode(value.toLong())

    /**
     * Decodes a Varint from a byte array starting at the given offset.
     *
     * @param bytes The byte array to read from.
     * @param offset Starting position in the array.
     * @return A [DecodeResult] containing the decoded value and the number of bytes consumed.
     * @throws IllegalArgumentException if the varint is malformed.
     */
    fun decode(bytes: ByteArray, offset: Int = 0): DecodeResult {
        var result = 0L
        var shift = 0
        var pos = offset

        while (pos < bytes.size) {
            val byte = bytes[pos].toInt() and 0xFF
            result = result or ((byte.toLong() and 0x7F) shl shift)
            pos++

            if (byte and 0x80 == 0) {
                return DecodeResult(result, pos - offset)
            }

            shift += 7
            require(shift < 64) { "Varint too long (more than 10 bytes)" }
        }

        throw IllegalArgumentException("Unexpected end of varint at offset $offset")
    }

    private fun encodeFull(value: Long): ByteArray {
        val bytes = ByteArray(10)
        var remaining = value
        for (i in 0 until 10) {
            bytes[i] = ((remaining and 0x7F) or (if (i < 9) 0x80 else 0)).toByte()
            remaining = remaining ushr 7
        }
        return bytes
    }

    /**
     * Result of decoding a Varint.
     *
     * @property value The decoded unsigned integer value.
     * @property bytesRead Number of bytes consumed from the input.
     */
    data class DecodeResult(val value: Long, val bytesRead: Int) {
        /** Convenience accessor for Int values. */
        val intValue: Int get() = value.toInt()
    }
}
